### DISCUSSION TITLE
Numpy and Tensorflow lite libraries for M5Stack ESP32 device

### DISCUSSION THREAD

**Discussion by MaxRondelli** [micropython/micropython#15773] [ESP32]
Hi there, 

I am new with micropython development and I am struggling to have the correct libraries for my project. I need at least Numpy and TensorFlow Lite. I have to run inference for a LSTM model with IMU data on the M5Stack ESP32 device. 

I've seen online that ulab is the corresponding numpy for micropython. I have tried to flash this [firmware](https://gitlab.com/rcolistete/micropython-firmwares/-/blob/master/ESP32/v1.12_with_ulab/ulab_v0.54.0_2020-07-29/Generic_flash-4MB/esp32_idf3_ulab_dp_thread_v1.12-663-gea4670d5a_2020-07-29.bin) but it doesn't work. I didn't find any solutions yet. 

Do you have any suggestion/tutorial on how I can download the correct firmware both for ulab and tflite?

Thank you.

Max

**Comment by bidrohini** [comment]
You can try TensorFlow Lite for Microcontrollers:
https://www.tensorflow.org/lite/microcontrollers

**Reply by MaxRondelli** [reply]
I've seen it thank you. 

I am struggling with the deploy of lib on my M5Stack device, in order to interpreter the `.tflite` model.
If you have any specific doc about it, I would be very grateful to you.

**Reply by sophiedegran** [reply]
Hi Max
Did you have a look how to compil micropython with ulab ?
github.com/v923z/micropython-ulab
You have first to determine which version of micropython works with your M5Stack (better to have last stable) after that which component of ulab you want to install.. 
It's work very well on Esp32S3 ;)

**Reply by MaxRondelli** [reply]
Hi @sophiedegran, yes I've solved my issue. I found out how to install ulab on my M5Stack device. Thank you.

**Reply by bidrohini** [reply]
In case someone else looks for ulab documentations in the future, here it goes.
https://docs.circuitpython.org/en/latest/shared-bindings/ulab/index.html

**Reply by scruss** [reply]
> In case someone else looks for ulab documentations in the future, here it goes. https://docs.circuitpython.org/en/latest/shared-bindings/ulab/index.html

That's the CircuitPython version. Here is the MicroPython one: [Welcome to the ulab book!](https://micropython-ulab.readthedocs.io/en/stable/)

**Comment by 2dof** [comment]
If You will not resolve Your problem with  tensorflow (Lite) package for your device, I would suggest to implement LSTM from scrach ( best with ulab) at your device - it require some coding and testing ( just rewriting code from python)  but at least You will get LSTM on Your devise.
see : [LSTM Neural Network from Scratch](https://www.kaggle.com/code/navjindervirdee/lstm-neural-network-from-scratch)
